### ISSUE TITLE
Data channel can't send or receive ArrayBuffer's

### ISSUE THREAD

**Issue by stefhak** [EricssonResearch/openwebrtc#287]
It seems that the current data channel implementation cannot handle ArrayBuffers. When sent to another browser (i.e. Firefox), what is delivered to the application on the receiving end is an empty object. When another browser sends an ArrayBuffer no onmessage ever fires.

**Comment by szimek** [comment]
Any idea when this feature might be implemented? It looks like there's no other way to send large amounts of data: `FileReader#readAsBinaryString` is deprecated according to MDN, `readAsDataURL` is not very efficient and sending blobs [is not supported by Chrome yet](https://code.google.com/p/webrtc/issues/detail?id=2276) and it looks like it won't be for a while.

**Comment by Rugvip** [comment]
I took a shot at implementing this, and it turns out it requires some changes to the c API or Seed. The blocking issue is that Seed isn't able to marshal the arguments of the on-binary-data signal, so it's not possible to receive messages atm. The solutions that I can think of are to either implement marshalling of arrays in signals in seed, or to switch to a boxed type e.g. GByteArray or GBytes.

I did implement sending ArrayBuffers and ArrayBufferViews though, might create a PR for that.

**Comment by stefhak** [comment]
@Rugvip : For clarification: are you saying that you can (with your fix) _send_ ArrayBuffers, but not recieive? It would be interesting anyhow, a first test (with @szimek's app perhaps) to send from owr and receive on FF would be a good start.

**Comment by Rugvip** [comment]
@stefhak Yep, it should be possible to send ArrayBuffers and ArrayBufferViews

**Comment by stefhak** [comment]
@Rugvip Could you supply your fix in a way that allows me to apply it and test?

**Comment by Rugvip** [comment]
@stefhak This should fetch the changes from my PR:
```
git checkout -b Rugvip-dc-binary master
git pull https://github.com/Rugvip/openwebrtc.git dc-binary
```

**Comment by stefhak** [comment]
@Rugvip Thanks. I applied the change and re-built, and it works for med sending ArrayBuffers from Safari (bakced by OWR) to Firefox, so it seems to be in order.

**Comment by superdump** [comment]
I've merged the sending part, so it's just the ArrayBuffer handling in Seed or in our annotations or whatever that needs fixing to be able to handle receiving them.

**Comment by superdump** [comment]
Changing title to can't receive.

**Comment by szimek** [comment]
Any updates?

**Comment by stefhak** [comment]
@szimek I'm afraid little has happened on this one :-( Is there any chance you could take a look and perhaps do a PR?

**Comment by roojs** [comment]
Hi, Got a message from Simon Hossell about the Issue, I maintain seed (albeit mostly in maintenance mode) 

I did look at the ArrayBuffer and file-io issue a few years ago, basically from what I recall, it's not feasible to access the internal ArrayBuffer implementation directly in Webkit/JScore

The only access would be calling the javascript API with char's, to build or fetch, which would not be at all efficient, actually I did end up doing something similar fetching binary files via webkit and writing to disk via stream.put_byte() - nasty but performance was not an issue for that usage.

To make this work basically you would need to create either a simple class in an extension that would implement a 'look-a-like' ArrayBuffer, or implement the specific code you need as a module function in C.

feel free to email me direct if you need anything alan@roojs.com - I'm not sure how well notifications work for this issue to my inbox..
